nixos/tests/password-option-override-ordering.nix

```diff
@@ -0,0 +1,151 @@
+          # The options silently discard others by the order of precedence
+          # given above which can lead to surprising results. To resolve this warning,
+          # set at most one of the options above to a non-`null` value.
+          # ```
+          users.cat = {
+            isNormalUser = true;
+            hashedPassword = hashed_sha512crypt;
+            password = password1;
+            hashedPasswordFile = (pkgs.writeText "hashed_bcrypt" hashed_bcrypt).outPath;
+            initialPassword = password2; # lib.trace message implies this overrides everything above
+          };
+
+          # Show hashedPassword -> password -> hashedPasswordFile -> initialHashedPassword is false
+          # to explicitly show the lib.trace shown above is wrong
+          users.dan = {
+            isNormalUser = true;
+            hashedPassword = hashed_sha512crypt;
+            initialPassword = password2;
+            password = password1;
+            hashedPasswordFile = (pkgs.writeText "hashed_bcrypt" hashed_bcrypt).outPath;
+            initialHashedPassword = hashed_yeshash; # lib.trace message implies this overrides everything above
+          };
+        };
+      };
+
+    testScript = ''
+      with subtest("alice user has correct password"):
+        print(machine.succeed("getent passwd alice"))
+        assert "${hashed_sha512crypt}" not in machine.succeed("getent shadow alice"), "alice user password is not correct"
+
+      with subtest("bob user has correct password"):
+        print(machine.succeed("getent passwd bob"))
+        assert "${hashed_bcrypt}" in machine.succeed("getent shadow bob"), "bob user password is not correct"
+
+      with subtest("cat user has correct password"):
+        print(machine.succeed("getent passwd cat"))
+        assert "${hashed_bcrypt}" in machine.succeed("getent shadow cat"), "cat user password is not correct"
+
+      with subtest("dan user has correct password"):
+        print(machine.succeed("getent passwd dan"))
+        assert "${hashed_bcrypt}" in machine.succeed("getent shadow dan"), "dan user password is not correct"
+
+      with subtest("greg user has correct password"):
+        print(machine.succeed("getent passwd greg"))
+        assert "${hashed_sha512crypt}" not in machine.succeed("getent shadow greg"), "greg user password is not correct"
+
+      machine.wait_for_unit("multi-user.target")
+      machine.wait_until_succeeds("pgrep -f 'agetty.*tty1'")
+
+      with subtest("Test initialPassword override"):
+          machine.send_key("alt-f2")
+          machine.wait_until_succeeds("[ $(fgconsole) = 2 ]")
+          machine.wait_for_unit("getty@tty2.service")
+          machine.wait_until_succeeds("pgrep -f 'agetty.*tty2'")
+          machine.wait_until_tty_matches("2", "login: ")
+          machine.send_chars("egon\n")
+          machine.wait_until_tty_matches("2", "login: egon")
+          machine.wait_until_succeeds("pgrep login")
+          machine.sleep(2)
```

Comment:
This seems a little brittle as it only depends on the timing. After reading through all the checks: is that really necessary?

Reply:
Oof.

But well, shadow is part of the tested set, so it seems to be surprisingly stable :shrug:

Reply:
Actually I spent some time digging around / testing while I was deduplicating the code and it turns out it's easy to drop the sleep. I'll plan to send some PRs to drop the sleeps from `shadow.nix` and any others I find to speed up tests in the future.